Allow restriction of the "Key" value in a Key/Value field to a white list of acceptable keys

**Is your feature request related to a problem? Please describe.**

The Key/Value field allows addition of multiple values to a content item within a single field, which is a commonly needed feature.  However as it is, the field allows content editors to enter any value for the key - which means that different content items may end up with completely different keys that are really meant to hold the same type of information.

**Describe the solution you'd like**

Allow specification of a list of acceptable key values for any Key/Value field.  For example, for a K/V field intended to store laptop specifications, you would be able to enforce that content editors can add keys for "Screen Resolution", "Power Consumption", "Screen Size", etc. - but prevent them from adding a key which is not in the list.

**Describe alternatives you've considered**

There are a couple of possible alternatives to this that are sometimes used:
* Use a relationship, where each key/value pair is entered as a different content item, and then related via a 1-N or M-N relationship.
* Use a custom field - where the customer has to write all the code to display the keys, allow the keys to be entered, encode the keys and values, etc.

**Additional context**

It would be ideal if, with a whitelist of keys specified, the user was given a drop-down list to choose the key from (but one they can type into, which will automatically match valid keys based on what they type).

It would also be ideal if, even if the keys are not restricted, entering keys behaved almost the same as entering tags - including autocomplete of previously used keys and/or values.  I think this is a heavier lift, though, than just restricting the keys to a list of acceptable values.

Looks like storing a json in a field variable Breaks the content-edit-window
with the following stack trace:

`Caused by: org.apache.jasper.JasperException: An exception occurred processing [/html/portlet/ext/contentlet/field/edit_field.jsp] at line [1118]

1115:         }
1116:
1117:         keyValueDataRaw = keyValueDataRaw.substring(0, keyValueDataRaw.length() - 1) + "}";
1118:         dotKeyValueDataRaw = dotKeyValueDataRaw.substring(0, dotKeyValueDataRaw.length() - 1);
1119:
1120:         List<FieldVariable> fieldVariables=APILocator.getFieldAPI().getFieldVariablesForField(field.getInode(), user, true);
1121:         String whiteListKeyValues = "";
`

--- a/dotCMS/src/main/webapp/html/portlet/ext/contentlet/field/edit_field.jsp
+++ b/dotCMS/src/main/webapp/html/portlet/ext/contentlet/field/edit_field.jsp
@@ -1106,16 +1106,25 @@
             keyValueMap.put("content", "...");
         }
 
-        String keyValueDataRaw = "{";
-        String dotKeyValueDataRaw = "";
+        final StringBuilder keyValueDataRaw = new StringBuilder("{");
+        final StringBuilder dotKeyValueDataRaw = new StringBuilder();
 
-        for(String key : keyValueMap.keySet()) {
-            keyValueDataRaw += key + ":" + UtilMethods.htmlifyString(UtilMethods.escapeDoubleQuotes(keyValueMap.get(key).toString())) + ",";
-            dotKeyValueDataRaw += key + "|" + UtilMethods.htmlifyString(UtilMethods.escapeDoubleQuotes(keyValueMap.get(key).toString())) + ",";
+        final Iterator<String> iterator = keyValueMap.keySet().iterator();
+
+        while (iterator.hasNext()) {
+            final String key = iterator.next();
+            final Object object = keyValueMap.get(key);
+            if(null != object) {
+                final String sanitized = UtilMethods.htmlifyString(UtilMethods.escapeDoubleQuotes(object.toString()));
+                keyValueDataRaw.append(key).append(":").append(sanitized);
+                dotKeyValueDataRaw.append(key).append("|").append(sanitized);
+                if (iterator.hasNext()) {
+                    keyValueDataRaw.append(',');
+                    dotKeyValueDataRaw.append(',');
+                }
+            }
         }
-
-        keyValueDataRaw = keyValueDataRaw.substring(0, keyValueDataRaw.length() - 1) + "}";
-        dotKeyValueDataRaw = dotKeyValueDataRaw.substring(0, dotKeyValueDataRaw.length() - 1);
+        keyValueDataRaw.append('}');
 
         List<FieldVariable> fieldVariables=APILocator.getFieldAPI().getFieldVariablesForField(field.getInode(), user, true);
         String whiteListKeyValues = "";
@@ -1125,7 +1134,7 @@
             }
         }
     %>
-        <input type="hidden" class ="<%=field.getVelocityVarName()%>" name="<%=field.getFieldContentlet()%>" id="<%=field.getVelocityVarName()%>" value="<%=keyValueDataRaw%>" />
+        <input type="hidden" class ="<%=field.getVelocityVarName()%>" name="<%=field.getFieldContentlet()%>" id="<%=field.getVelocityVarName()%>" value="<%=keyValueDataRaw.toString()%>" />
         <style>
             dot-key-value key-value-table tr {
                 cursor: move;
@@ -1188,7 +1197,7 @@
         <script>
             var dotKeyValue = document.querySelector('dot-key-value');
             dotKeyValue.uniqueKeys = "true";
-            dotKeyValue.value = '<%=dotKeyValueDataRaw%>';
+            dotKeyValue.value = '<%=dotKeyValueDataRaw.toString()%>';
             dotKeyValue.disabled = '<%=field.isReadOnly()%>';
             dotKeyValue.whiteList = '<%=whiteListKeyValues%>';
             dotKeyValue.formKeyLabel = '<%= LanguageUtil.get(pageContext, "Key") %>'

